File: cs/algorithms_in/insertion_sort.cs
namespace algorithms_in;

public class InsertionSort
{
    public static void insertionSort(int[] array) {
        for (int i = 1; i < array.Length; i++) {
            int currentElement = array[i];
            int k;
            for (k = i - 1; k >= 0 && array[k] > currentElement; k--) {
                array[k + 1] = array[k];
            }
            array[k + 1] = currentElement;
        }
    }

    public static void insertionSort2(int[] array) {
        for (int i = 1; i < array.Length; i++) {
            for (int k = i; k > 0 && array[k - 1] > array[k]; k--) {
                int temp = array[k - 1];
                array[k - 1] = array[k];
                array[k] = temp;
            }
        }
    }
}

File: rust/algorithms_in/src/insertion_sort.rs
pub fn insertion_sort(array: &mut [usize]) {
    for i in 1..array.len() {
        let mut k = i;
        while k > 0 && array[k - 1] > array[k] {
            array.swap(k, k - 1);
            k -= 1;
        }
    }
}

pub fn insertion_sort_2(array: &mut [usize]) {
    for i in 1..array.len() {
        for k in (0..i).rev() {
            if array[k] >= array[k + 1] {
                array.swap(k, k + 1);
            } else {
                break;
            }
        }
    }
}